packages/taquito/src/prepare/prepare-provider.ts

```diff
@@ -1092,4 +1098,96 @@ export class PrepareProvider extends Provider implements PreparationProvider {
       contents,
     };
   }
+
+  /**
+   *
+   * @description Method to prepare a stake operation
+   * @param params RPCOperation object
+   * @param source string or undefined source pkh
+   * @returns a PreparedOperation object
+   */
+  stake(params: StakeParams): Promise<PreparedOperation> {
+    return this.staking({
+      ...params,
+      parameter: {
+        entrypoint: 'stake',
+        value: { prim: 'Unit' },
+      },
+    });
+  }
+
+  /**
+   *
+   * @description Method to prepare an unstake operation
+   * @param params RPCOperation object
+   * @param source string or undefined source pkh
+   * @returns a PreparedOperation object
+   */
+  async unstake({ amount, mutez, ...rest }: UnstakeParams): Promise<PreparedOperation> {
+    return this.staking({
+      parameter: {
+        entrypoint: 'unstake',
+        value: { int: mutez ? amount.toString() : format('tz', 'mutez', amount).toString() },
+      },
+      ...rest,
+      amount: 0,
+    });
+  }
+
+  /**
+   *
+   * @description Method to prepare a finalize_unstake operation
+   * @param params RPCOperation object
+   * @returns a PreparedOperation object
+   */
+  async finalizeUnstake(params: FinalizeUnstakeParams): Promise<PreparedOperation> {
+    return this.staking({
+      parameter: {
+        entrypoint: 'finalize_unstake',
+        value: { prim: 'Unit' },
+      },
+      ...params,
+      amount: 0,
+    });
+  }
+
+  private async staking({
+    fee,
+    storageLimit,
+    gasLimit,
+    source,
+    ...rest
+  }: StakingParams): Promise<PreparedOperation> {
+    const { pkh } = await this.getKeys();
+
+    const protocolConstants = await this.context.readProvider.getProtocolConstants('head');
+    const DEFAULT_PARAMS = await this.getAccountLimits(pkh, protocolConstants);
+    const mergedEstimates = mergeLimits({ fee, storageLimit, gasLimit }, DEFAULT_PARAMS);
+
+    const op = await createStakingOperation({
+      source: source ?? pkh,
+      ...rest,
+      ...mergedEstimates,
+    });
+
+    const operation = await this.addRevealOperationIfNeeded(op, pkh);
+    const ops = this.convertIntoArray(operation);
+
+    const hash = await this.getBlockHash();
+    const protocol = await this.getProtocolHash();
+
+    this.#counters = {};
+    const headCounter = parseInt(await this.getHeadCounter(pkh), 10);
+
+    const contents = this.constructOpContents(ops, headCounter, pkh, source ?? pkh);
```

Comment:
feels unnecessary to have ` ?? pkh`

Reply:
they put source and pkh separately so contracutOpContents can have source to fall on. And inside contracutOpContents will now have circular logic